ledger/apptxn_test.go

```diff
@@ -0,0 +1,418 @@
+	eval.txn(t, withdraw.Noted("2"))
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in) // Zero left, but still opted in
+	require.Equal(t, amount, uint64(0))
+
+	eval = l.nextBlock(t)
+	eval.txn(t, withdraw.Noted("3"), "underflow on subtracting")
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.True(t, in) // Zero left, but still opted in
+	require.Equal(t, amount, uint64(0))
+
+	close := txntest.Txn{
+		Type:            "appl",
+		ApplicationID:   appIndex,
+		Sender:          addrs[0],
+		ApplicationArgs: [][]byte{[]byte("close")},
+		ForeignAssets:   []basics.AssetIndex{asaIndex},
+		Accounts:        []basics.Address{addrs[0]},
+	}
+
+	eval = l.nextBlock(t)
+	eval.txn(t, &close)
+	l.endBlock(t, eval)
+
+	amount, in = l.asa(t, appIndex.Address(), asaIndex)
+	require.False(t, in) // Zero left, not opted in
+	require.Equal(t, amount, uint64(0))
+
+	// Now, fail again, opted out
+	eval = l.nextBlock(t)
+	eval.txn(t, fundgold.Noted("2"), fmt.Sprintf("asset %d missing", asaIndex))
+	l.endBlock(t, eval)
+}
+
+// TestClawbackAction ensures an app address can act as clawback address.
+func TestClawbackAction(t *testing.T) {
+	genBalances, addrs, _ := newTestGenesis()
+	l := newTestLedger(t, genBalances)
+	defer l.Close()
+
+	// Would be better to pull these out of block, or at least check them.
+	asaIndex := basics.AssetIndex(1)
+	appIndex := basics.AppIndex(2)
+
+	asa := txntest.Txn{
+		Type:   "acfg",
+		Sender: addrs[0],
+		AssetParams: basics.AssetParams{
+			Total:     1000000,
+			Decimals:  3,
+			UnitName:  "oz",
+			AssetName: "Gold",
+			URL:       "https://gold.rush/",
+			Clawback:  appIndex.Address(),
+		},
+	}
+
+	app := txntest.Txn{
+		Type:   "appl",
+		Sender: addrs[0],
+		ApprovalProgram: main(`
+         tx_begin
+
+         int axfer
+         tx_field TypeEnum
+
+         txn Assets 0
+         tx_field XferAsset
+
+         txn Accounts 1
+         tx_field AssetSender
+
+         txn Accounts 2
+         tx_field AssetReceiver
+
+         int 1000
+         tx_field AssetAmount
+
+         tx_submit
+`),
+	}
+
+	optin := txntest.Txn{
+		Type:          "axfer",
+		Sender:        addrs[1],
+		AssetReceiver: addrs[1],
+		XferAsset:     asaIndex,
+	}
+	eval := l.nextBlock(t)
+	eval.txns(t, &asa, &app, &optin)
+	l.endBlock(t, eval)
+
+	bystander := addrs[2] // Has no authority of its own
+	overpay := txntest.Txn{
+		Type:     "pay",
+		Sender:   bystander,
+		Receiver: bystander,
+		Fee:      2000, // Overpay fee so that app account can be unfunded
+	}
+	clawmove := txntest.Txn{
+		Type:          "appl",
+		Sender:        bystander,
+		ApplicationID: appIndex,
+		ForeignAssets: []basics.AssetIndex{asaIndex},
+		Accounts:      []basics.Address{addrs[0], addrs[1]},
+	}
+	eval = l.nextBlock(t)
+	eval.txgroup(t, &overpay, &clawmove)
+	l.endBlock(t, eval)
+
+	amount, _ := l.asa(t, addrs[1], asaIndex)
+	require.Equal(t, amount, uint64(1000))
```

Comment:
let's test txn counter in a block header as well - to double check it gets incremented